beps/0012-metrics-service/README.md

```diff
@@ -0,0 +1,255 @@
+
+```ts
+interface MetricsService {
+  // Synchronous instrumentation
+  createCounter(name: string, options?: MetricOptions): Counter;
+  createUpDownCounter(name: string, options?: MetricOptions): UpDownCounter;
+  createHistogram(name: string, options?: MetricOptions): Histogram;
+  createGauge(name: string, options?: MetricOptions): Gauge;
+
+  // Asynchronous instrumentation
+  createObservableCounter(
+    name: string,
+    options?: MetricOptions,
+  ): ObservableCounter;
+  createObservableUpDownCounter(
+    name: string,
+    options?: MetricOptions,
+  ): ObservableUpDownCounter;
+  createObservableGauge(name: string, options?: MetricOptions): ObservableGauge;
+
+  // Provide access to the underlying meter for advanced configuration.
+  getMeter(): Meter;
+
+  // Additional convenience methods tbd...
+}
+```
+
+The `RootMetricsService` is responsible for the global meter, namespaced to `backstage` or the `serviceName` and `serviceVersion` provided in the app's `app-config.yaml` file. The root metrics service will depend on as little as possible so that it can be initialized as early as possible.
+
+```ts
+export const rootMetricsServiceFactory = createServiceFactory({
+  service: rootMetricsServiceRef,
+  deps: {
+    rootConfig: coreServices.rootConfig,
+  },
+  factory: ({ rootConfig }) => {
+    return DefaultRootMetricsService.fromConfig(rootConfig);
+  },
+});
+```
+
+It also provides a method for creating a plugin-scoped `MetricsService` for each registered plugin.
+
+```ts
+interface RootMetricsService extends MetricsService {
+  forPlugin(pluginId: string): MetricsService;
+}
+
+class DefaultRootMetricsService {
+  // ...
+
+  static fromConfig(config: Config): RootMetricsService {
+    return new DefaultRootMetricsService(config);
+  }
+
+  forPlugin(pluginId: string): MetricsService {
+    return new PluginMetricsService({
+      pluginId,
+      serviceName: this.serviceName,
+      serviceVersion: this.serviceVersion,
+    });
+  }
+}
+```
+
+Each plugin-scoped `MetricsService` acquires a meter when it is created by calling `metrics.getMeter` from the `@opentelemetry/api` package. The internals of the service ensure the metric name is namespaced to match the naming conventions outlined in the [naming conventions](#naming-conventions) section.
+
+```ts
+class PluginMetricsService implements MetricsService {
+  // ...
+  constructor({
+    pluginId,
+    serviceName,
+    serviceVersion,
+  }: PluginMetricsServiceOptions) {
+    this.pluginId = pluginId;
+    this.serviceName = serviceName;
+    this.meter = metrics.getMeter(serviceName, serviceVersion);
+  }
+
+  // Ensures consistency across all metrics
+  private prefixMetricName(name: string): string {
+    return `${this.serviceName}.plugin.${this.pluginId}.${name}`;
+  }
+}
+```
+
+### Naming Conventions
+
+All Backstage metrics MUST follow this hierarchical pattern:
+
+`backstage.{scope}.{scope_name}.{category}.{metric_name}`
```

Comment:
How do we plan to handle the existing metrics, will those stay for some period with a deprecation wanting before being removed? None of them match this pattern: https://backstage.io/docs/tutorials/setup-opentelemetry#available-metrics. If this is the plan we should also have some way to note deprecated metrics in the logs as well.